### DISCUSSION TITLE
[REQUEST]: Allow skipping environment approval on specific branches

### DISCUSSION THREAD

**Discussion by maxnorth** [community/community#112177] [Actions]
### Select Topic Area

Product Feedback

### Body

We'd like to be able to require manual approval to our staging environment from feature branches, but have them automatically proceed when running from the `main` branch. We allow developers to optionally test in staging from their feature branches, but we want every main commit to go to staging as part of our continuous delivery pipeline. 

Currently I see no secure way to do this. 

I see a couple options to get the desired workflow, but both undermine security and cause other maintenance headaches:
1) Create two 'environment' definitions for staging, one with and one without the approval condition. This is bad for maintainability because it requires manual duplication/synchronization of variables and secrets between environments. It's bad for security, because any contributor could easily swap the environment values and run staging deployments without the approval from feature branches. It's also a clear hack that undermines the design of the "environment" concept in Github actions. We don't have two staging environments, so the two instances would no longer represent environments, they'd just be functional needs imposed on the environment feature.
2) Another option that I've seen suggested more frequently in here is to store a PAT as a repo secret that has permissions to do approvals, and conditionally send a REST request within the workflow. This fully undermines the security goals of environments, as anyone could use that PAT to approve any environment that the PAT's user is an approver for. We also avoid using PAT's in workflows as they depend on user's inclusion in the org, which will unexpectedly break in a few years whenever that person leaves the company.

Allowing environments to be configured to bypass approvals for specific branches seems like the simplest and most desirable approach. This still enables the repo admins to control the conditions under which environment access can occur, since they can set branch protection rules that must be met to merge to that branch.

**Comment by maxnorth** [comment]
A partial self-correction - I realized that in option 1, you can avoid a security vulnerability if you apply a branch requirement on the version of the environment that does not have an approval requirement. That would prevent workflow authors from swapping them. However, the points about redundant environment variable/secret management and clunky semantics of what an "environment" is at that point still stand.

It's interesting that branch conditions are already supported, but can only be combined with other conditions like approvals using "AND" logic. Really the functionality needed here is just to support "OR" conditions for multiple requirements.

**Comment by AlistairGempf** [comment]
Just wanted to add an agreement to this as it's not been replied/actioned in over a year. We require approvals from another dev before merge to `main`, so we implicitly already have approval to deploy to our first environment on merge. We also want to be able to test from feature branches on the first environment, but would like that to be approved by another developer. An "OR" condition on the environment would make our workflows much simpler.

I _think_ a newish third option is using a [custom deployment rule](https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-deployments/creating-custom-deployment-protection-rules), but that feels to me like using a hydraulic press to crack a nut and has trust and security implications depending on the service you're using to approve the deployment.

**Reply by Richiban** [reply]
This is exactly our use case as well: our `main` branch is protected so we don't need deployment approval for the main branch but we *do* want it for all other branches. As far as I can see there's no workaround for this; we thought of having a dedicated `test` branch that also requires an approved PR but we can't think of any way to "reset" the `test` branch once someone's done with it.
